### PR TITLE
feat(auth): add confirmUserAttribute API

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

confirmUserAttribute allows to confirm a user attribute while the user is authenticated

**API changes:**

- added category API types
- added API inside cognito directory with cognito type definitions

**Other changes**
- addressed @Samaritan1011001  feedback from previous PR to change the name of the following helper functions: `getUpdatedAttributes` -> `getConfirmedAttributes` , `getUnUpdatedAttributes` -> `getUnConfirmedAttributes`

**Testing**
- tests that the api calls the service
- test validation error when `confirmationCode` is empty
- tests service error

#### Description of how you validated changes

Tested API using a sample-app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by Samaritan1011001 — DISMISSED** [review]
LGTM

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import { AuthValidationErrorCode } from '../../../errors/types/validation';
+import { assertValidationError } from '../../../errors/utils/assertValidationError';
+import { ConfirmUserAttributeRequest } from '../../../types/requests';
+import { verifyUserAttribute } from '../utils/clients/CognitoIdentityProvider';
+import { VerifyUserAttributeException } from '../types/errors';
+import {
+	AmplifyV6 as Amplify,
+	assertTokenProviderConfig,
+} from '@aws-amplify/core';
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:12]: Nit: Would prefer for absolute imports to be at the top. Also I think `assertTokenProviderConfig` needs to be imported from `@aws-amplify/core/internals/utils` due to a recent export change.

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import { fetchAuthSession } from '../../../';
+import { getRegion } from '../utils/clients/CognitoIdentityProvider/utils';
+import { assertAuthTokens } from '../utils/types';
+import { CognitoUserAttributeKey } from '../types';
+
+/**
+ * Confirm the user attribute with the confirmation code.
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:19]:
```suggestion
 * Confirms a user attribute with the confirmation code.
```

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @param confirmUserAttributeRequest - The ConfirmUserAttributeRequest
+ *
+ * @throws  -{@link AuthValidationErrorCode } -
+ * Thrown when `confirmationCode` is not defined.
+ *
+ * @throws  -{@link VerifyUserAttributeException }
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:26]: Should this have a description?

```suggestion
 * @throws  -{@link VerifyUserAttributeException }
```

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @throws AuthTokenConfigException - Thrown when the token provider config is invalid.
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:28]: Format seems different than the others

```suggestion
 * @throws AuthTokenConfigException - Thrown when the token provider config is invalid.
```

**Review by jimblanc — COMMENTED** [review]
Looks good! Some nits & comments to fix build

**Review by jimblanc — APPROVED** [review]
LGTM

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	confirmUserAttributeRequest: ConfirmUserAttributeRequest<CognitoUserAttributeKey>
+): Promise<void> {
+	const authConfig = Amplify.getConfig().Auth;
+	assertTokenProviderConfig(authConfig);
+	const { confirmationCode, userAttributeKey } = confirmUserAttributeRequest;
+	assertValidationError(
+		!!confirmationCode,
+		AuthValidationErrorCode.EmptyConfirmUserAttributeCode
+	);
+	const { tokens } = await fetchAuthSession({ forceRefresh: false });
```
Comment (Samaritan1011001) [inline packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:38]: nit: but isn't `forceRefresh` `false` by default?

**Review by Samaritan1011001 — APPROVED** [review]
lgtm, just a small nit

File: packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts (hunk 5 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/confirmUserAttribute.ts:38]: yeah that's true. The other option is to call the API with an empty object. I don't see it to be explicit. Seems pretty confusing.

**Pull request merged**